lib/Service/RelatedService.php

```diff
@@ -120,104 +127,114 @@ public function getRelatedToItem(string $providerId, string $itemId, int $chunk
-							}
-						}
-					}
-
-					// improve score on duplicate result
-					if (in_array($related->getItemId(), $known)) {
-						try {
-							$knownRecipient = $this->extractRecipientFromResult(
-								$related->getProviderId(),
-								$related->getItemId(),
-								$result
-							);
+			foreach ($recipients as $recipient) {
+				// foreach provider, we get a list of items available to each recipient of the 'current' item
+				// we only needs itemIds because, at this point, the full list of recipient each
+				// item is shared to is not important
+				// However, if 'current' item contains a group share, we do not need to waste resource to get
+				// details about items available to single users as they are ignored in current scope of the app.
+				try {
+					$entity = $this->circlesManager->getFederatedUser($recipient);
+				} catch (Exception $e) {
+					continue;
+				}
 
-							$knownRecipient->improve(RelatedResource::$IMPROVE_OCCURRENCE, 'occurrence');
-						} catch (ItemNotFoundException $e) {
-						}
+				if ($current->isGroupShared() && $entity->getBasedOn()->getSource() === Member::TYPE_USER) {
+					continue;
+				}
 
-						continue;
+				foreach ($this->getItemsAvailableToEntity($provider, $entity) as $itemId) {
+					if (in_array($itemId, $known)) {
+						continue; // we don't want duplicate details
 					}
+					$known[] = $itemId;
 
-					if ($provider->getProviderId() !== $providerId || $related->getItemId() !== $itemId) {
-						$result[] = $related;
+					// foreach itemId, we get full details about it
+					try {
+						// cast to string is mandatory in here !
+						$related = $this->getRelatedFromItem($provider->getProviderId(), (string)$itemId);
+					} catch (RelatedResourceNotFound $e) {
+						continue;
 					}
 
-					$known[] = $related->getItemId();
+					$result[] = $related;
 				}
 			}
 		}
 
+		$result = $this->strictMatching($current, $result);
 		$result = $this->filterUnavailableResults($result);
+
 		if (!empty($itemPaths)) {
 			$this->weightResult($itemPaths, $result);
 		}
 
-		return $this->filterLowScoreResults($result);
+		return $result;
 	}
 
 
 	/**
+	 * get the RelatedResource from an item. including all recipient/virtual groups
+	 *
 	 * @param string $providerId
 	 * @param string $itemId
 	 *
-	 * @return FederatedUser[]
+	 * @return RelatedResource
+	 * @throws RelatedResourceNotFound
 	 * @throws RelatedResourceProviderNotFound
 	 */
-	public function getSharesRecipients(string $providerId, string $itemId): array {
+	public function getRelatedFromItem(string $providerId, string $itemId): RelatedResource {
 		try {
-			$shares = $this->getCachedSharesRecipients($providerId, $itemId);
-
-			return $shares;
+			return $this->getCachedRelatedFromItem($providerId, $itemId);
 		} catch (CacheNotFoundException $e) {
 		}
 
+		$this->getRelatedResourceProvider($providerId);
 		$result = $this->getRelatedResourceProvider($providerId)
```

Comment:
```suggestion
		$result = $this->getRelatedResourceProvider($providerId)
```
Unless I misunderstood the first call is useless?